cf-check/diagnose.c

```diff
@@ -0,0 +1,353 @@
+    macro(SIGNAL_CONTINUE)                                \
+    macro(SIGNAL_CHILD_STATUS_CHANGE)                     \
+    macro(SIGNAL_BACKGROUND_READ_ATTEMPT)                 \
+    macro(SIGNAL_BACKGROUND_WRITE_ATTEMPT)                \
+    macro(SIGNAL_IO_POSSIBLE_ON_DESCRIPTOR)               \
+    macro(SIGNAL_CPU_TIME_EXCEEDED)                       \
+    macro(SIGNAL_FILE_SIZE_EXCEEDED)                      \
+    macro(SIGNAL_VIRTUAL_TIME_ALARM)                      \
+    macro(SIGNAL_PROFILING_TIMER_ALARM)                   \
+    macro(SIGNAL_WINDOW_SIZE_CHANGE)                      \
+    macro(SIGNAL_STATUS_REQUEST)                          \
+    macro(SIGNAL_OTHER)                                   \
+    macro(LMDB_KEY_EXISTS)                                \
+    macro(LMDB_KEY_NOT_FOUND)                             \
+    macro(LMDB_PAGE_NOT_FOUND)                            \
+    macro(LMDB_CORRUPT_PAGE)                              \
+    macro(LMDB_PANIC_FATAL_ERROR)                         \
+    macro(LMDB_VERSION_MISMATCH)                          \
+    macro(LMDB_INVALID_DATABASE)                          \
+    macro(LMDB_MAP_FULL)                                  \
+    macro(LMDB_DBS_FULL)                                  \
+    macro(LMDB_READERS_FULL)                              \
+    macro(LMDB_TLS_KEYS_FULL)                             \
+    macro(LMDB_TRANSACTION_FULL)                          \
+    macro(LMDB_CURSOR_STACK_TOO_DEEP)                     \
+    macro(LMDB_PAGE_FULL)                                 \
+    macro(LMDB_MAP_RESIZE_BEYOND_SIZE)                    \
+    macro(LMDB_INCOMPATIBLE_OPERATION)                    \
+    macro(LMDB_INVALID_REUSE_OF_READER_LOCKTABLE_SLOT)    \
+    macro(LMDB_BAD_OR_INVALID_TRANSACTION)                \
+    macro(LMDB_WRONG_KEY_OR_VALUE_SIZE)                   \
+    macro(LMDB_BAD_DBI)                                   \
+    macro(LMDUMP_UNKNOWN_ERROR)                           \
+    macro(PID_ERROR)                                      \
+    macro(PERMISSION_ERROR)                               \
+    macro(DOES_NOT_EXIST)                                 \
+    macro(UNKNOWN)
+
+#define CF_CHECK_MAX CF_CHECK_UNKNOWN
+
+#define CF_CHECK_CREATE_ENUM(name) \
+  CF_CHECK_##name,
+
+#define CF_CHECK_CREATE_STRING(name) \
+  #name,
+
+typedef enum {
+    CF_CHECK_RUN_CODES(CF_CHECK_CREATE_ENUM)
+} CFCheckCode;
+
+static const char *CF_CHECK_STR[] = {
+    CF_CHECK_RUN_CODES(CF_CHECK_CREATE_STRING)
+};
+
+static bool code_is_errno(int r)
+{
+    return (r > CF_CHECK_MAX);
+}
+
+// Better strerror, returns NULL if it doesn't know.
+static const char *strerror_or_null(int r)
+{
+    const char *strerror_string = strerror(r);
+    if (strerror_string == NULL)
+    {
+        return NULL;
+    }
+
+    const char *unknown = "Unknown error";
+    if (strncmp(strerror_string, unknown, strlen(unknown)) == 0)
```

Comment:
Not important here, but you could optimize this by doing:

```
const char unknown[] = "Unknown error";
if (strncmp(strerror_string, unknown, sizeof(unknown) - 1) == 0)
```

Reply:
Yes, that's neat.